From 9a2e60b60ce9791e12c482f15cd90c210e0ea9b2 Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Thu, 11 Sep 2014 22:43:39 -0400
Subject: [infiniband] cxgb4: Clean up connection on ARP error

Message-id: <1410475447-94392-81-git-send-email-svemuri@redhat.com>
Patchwork-id: 93829
O-Subject: [RHEL7.1 PATCH BZ 1124947 080/108] RDMA/cxgb4: Clean up connection on ARP error
Bugzilla: 1124947
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

Based on origninal work by Steve Wise <swise@opengridcomputing.com>

Signed-off-by: Steve Wise <swise@opengridcomputing.com>
Signed-off-by: Hariprasad Shenai <hariprasad@chelsio.com>
Signed-off-by: Roland Dreier <roland@purestorage.com>
(cherry picked from commit 5dab6d3ab1abed99be6166b844af58237d52a135)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/infiniband/hw/cxgb4/cm.c b/drivers/infiniband/hw/cxgb4/cm.c
index 4e59fd0..afb0058 100644
--- a/drivers/infiniband/hw/cxgb4/cm.c
+++ b/drivers/infiniband/hw/cxgb4/cm.c
@@ -432,8 +432,17 @@ static void arp_failure_discard(void *handle, struct sk_buff *skb)
  */
 static void act_open_req_arp_failure(void *handle, struct sk_buff *skb)
 {
+ struct c4iw_ep *ep = handle;
+
  printk(KERN_ERR MOD "ARP failure duing connect\n");
  kfree_skb(skb);
+ connect_reply_upcall(ep, -EHOSTUNREACH);
+ state_set(&ep->com, DEAD);
+ remove_handle(ep->com.dev, &ep->com.dev->atid_idr, ep->atid);
+ cxgb4_free_atid(ep->com.dev->rdev.lldi.tids, ep->atid);
+ dst_release(ep->dst);
+ cxgb4_l2t_release(ep->l2t);
+ c4iw_put_ep(&ep->com);
 }
 
 /*
@@ -659,7 +668,7 @@ static int send_connect(struct c4iw_ep *ep)
   opt2 |= T5_OPT_2_VALID;
   opt2 |= V_CONG_CNTRL(CONG_ALG_TAHOE);
  }
- t4_set_arp_err_handler(skb, NULL, act_open_req_arp_failure);
+ t4_set_arp_err_handler(skb, ep, act_open_req_arp_failure);
 
  if (is_t4(ep->com.dev->rdev.lldi.adapter_type)) {
   if (ep->com.remote_addr.ss_family == AF_INET) {
-- 
1.7.1